crates/ruff_linter/src/rules/pyupgrade/rules/pep695/use_pep695_type_parameter.rs

```diff
@@ -0,0 +1,237 @@
+                format!("Generic function `{name}` should use type parameters")
+            }
+        }
+    }
+
+    fn fix_title(&self) -> Option<String> {
+        Some("Use type parameters".to_string())
+    }
+}
+
+/// UP046
+pub(crate) fn non_pep695_generic_class(checker: &mut Checker, class_def: &StmtClassDef) {
+    if checker.settings.target_version < PythonVersion::Py312 {
+        return;
+    }
+
+    let StmtClassDef {
+        name,
+        type_params,
+        arguments,
+        ..
+    } = class_def;
+
+    // it's a runtime error to mix type_params and Generic, so bail out early if we see existing
+    // type_params
+    if type_params.is_some() {
+        return;
+    }
+
+    let Some(arguments) = arguments.as_ref() else {
+        return;
+    };
+
+    // TODO(brent) only accept a single, Generic argument for now. I think it should be fine to have
+    // other arguments, but this simplifies the fix just to delete the argument list for now
+    let [Expr::Subscript(ExprSubscript { value, slice, .. })] = arguments.args.as_ref() else {
+        return;
+    };
+
+    if !checker.semantic().match_typing_expr(value, "Generic") {
+        return;
+    }
+
+    let vars = {
+        let mut visitor = TypeVarReferenceVisitor {
+            vars: vec![],
+            semantic: checker.semantic(),
+        };
+        visitor.visit_expr(slice);
+        visitor.vars
+    };
+
+    // Type variables must be unique; filter while preserving order.
+    let mut type_vars = vars
+        .into_iter()
+        .unique_by(|TypeVar { name, .. }| name.id.as_str())
+        .collect::<Vec<_>>();
+
+    if type_vars.is_empty() {
+        return;
+    }
+
+    // generally preserve order, but sort by kind so that the order will be TypeVar...,
+    // TypeVarTuple..., ParamSpec...
+    type_vars.sort_by_key(|tv| tv.kind);
+
+    // build the fix as a String to avoid removing comments from the entire function body
+    let type_params = fmt_type_vars(&type_vars, checker);
+
+    checker.diagnostics.push(
+        Diagnostic::new(
+            NonPEP695TypeParameter {
+                name: name.to_string(),
+                generic_kind: GenericKind::GenericClass,
+            },
+            TextRange::new(name.start(), arguments.end()),
+        )
+        .with_fix(Fix::applicable_edit(
+            Edit::replacement(type_params, name.end(), arguments.end()),
+            Applicability::Safe,
+        )),
+    );
+}
+
+/// UP046
+pub(crate) fn non_pep695_generic_function(checker: &mut Checker, function_def: &StmtFunctionDef) {
+    if checker.settings.target_version < PythonVersion::Py312 {
+        return;
+    }
+
+    let StmtFunctionDef {
+        name,
+        type_params,
+        parameters,
+        ..
+    } = function_def;
+
+    // TODO(brent) handle methods, for now return early in a class body. For example, an additional
+    // generic parameter on the method needs to be handled separately from one already on the class
```

Comment:
Could you add a note to the rule's documentation saying that methods are currently not supported?